### ISSUE TITLE
Facilitate Map Packaging

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#98]
The how-to-release notes have a mention to download a number of maps to include them in the "maps" folder. This seems a bit hacky since some maps are there, and others are in another location.

I then suggest we do this, create two map folders, "maps" and "maps-extra"  (ideally the second would be called extra-maps, but to keep the two folders closer together when sorted, hence: map-extras)

Everything in maps would be packaged in the minimal version, used for engine upgrades and perhaps deployments. The maps-extra would be packaged in the 'all-maps' version and would include everything in 'maps' as well.

**Comment by DanVanAtta** [comment]
So there are two actions here, one is to split the current maps folders into two. Secondly, to import and check in the maps that are not yet in git but mentioned in the how to release wiki (like WWII 1941 for example)

**Comment by veqryn** [comment]
the 'extra maps' are stored in the other svn repo at the moment.  i do not like the idea of putting them into this git repo, because then they would be in two places.

**Comment by DanVanAtta** [comment]
My understanding is we wish to remove both SVN and sourceforge dependencies. Moving the maps to git would help accomplish that in addition to simplifying the release process for the moment (alternatively we script out some SVN checkout stuff). So in effect they would be in one place. I would suggest that they are in two places today since not all in SVN, nor all in Git.

**Comment by veqryn** [comment]
We've been through this one already: github can not handle the "tripleamaps" repo.

**Comment by DanVanAtta** [comment]
I'm not referring to the whole repo but the dozen maps you listed required to build.
https://github.com/triplea-game/triplea/wiki/_Release-Building-Artifacts

With respect to the whole repo, having another github repo and then adding each zip file as a release artifact probably could replace SVN if we desired. But that is a different topic. Point here is to simplify the release process, so we gain at least from that perspective that the files are in one place instead of two (from the SVN perspective they are already in two different places, so we lose nothing there)

**Comment by DanVanAtta** [comment]
"Optional in the future: switch to just using the extra map version, which users prefer because everyone comes for the WWIIvX games"

Just noticed the comment here. Not everyone comes for just the WWIIvX games. NWO, global, Big World v3 are the other popular games in the lobby. By that token, if players come for those games, why do we include middle earth, and 70M of napolean map skins with the download? This doesn't make sense to me. 

What I think would make more sense is a maps folder with the most played games, the ones listed above. Then the extra maps folder with Napolean, Middle Earth, etc.  Everything we would then package would be available in git.  (alternatively we update the build process to DL this stuff from SVN, which might be okay too, but we would still want the folder structure)

**Comment by gaborbernat** [comment]
`We've been through this one already: github can not handle the
"tripleamaps" repo.` -> Can you point me to this, why?

Bernát GÁBOR

**Comment by DanVanAtta** [comment]
IIRC, size mainly: 2.2G needed for all the maps unzipped, but 1GB github limit. (1.7GB needed for zips)

On the other hand, I've learned more about github releases since then. All we do is store the zip files in SVN, so it's essentially a version controlled file dump (which is okay). We suggest we use github releases for this same purpose, with a REST API that we can use for added benefit.

**Comment by gaborbernat** [comment]
why don't we just store each map in it's own repository?

On 8/20/15, Dan Van Atta <notifications@github.com> wrote:
> IIRC, size mainly: 2.2G needed for all the maps unzipped, but 1GB github
> limit. (1.7GB needed for zips)
>
> On the other hand, I've learned more about github releases since then. All
> we do is store the zip files in SVN, so it's essentially a version
> controlled file dump (which is okay). We suggest we use github releases for
> this same purpose, with a REST API that we can use for added benefit.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/triplea-game/triplea/issues/98#issuecomment-132911215


-- 
 Bernát Gábor
Student -  Budapest University of Technology and Economics - Computer
Engineering, M.Sc. - Budapest, Hungary
System Integrator - Gravity R&D | Rock Solid Recommendations - Budapest
office | 5-7 Expo ter | H-1101 | Budapest | Hungary

**Comment by DanVanAtta** [comment]
109 maps, lots of repo's. Even grouping to a dozen maps per repo is still a pretty significant number of repos.

**Comment by gaborbernat** [comment]
I would still go for it :) needs to be created only once :-? and a map
is sort of its own entity :) with its own releases :)

On 8/20/15, Dan Van Atta <notifications@github.com> wrote:
> 109 maps, lots of repo's. Even grouping to a dozen maps per repo is still a
> pretty significant number of repos.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/triplea-game/triplea/issues/98#issuecomment-132946529


-- 
 Bernát Gábor
Student -  Budapest University of Technology and Economics - Computer
Engineering, M.Sc. - Budapest, Hungary
System Integrator - Gravity R&D | Rock Solid Recommendations - Budapest
office | 5-7 Expo ter | H-1101 | Budapest | Hungary

**Comment by gaborbernat** [comment]
repos need to be created only once :)  we can make maps submodules of
the tripleA project and then could check out all at once if needed

On 8/20/15, Dan Van Atta <notifications@github.com> wrote:
> 109 maps, lots of repo's. Even grouping to a dozen maps per repo is still a
> pretty significant number of repos.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/triplea-game/triplea/issues/98#issuecomment-132946529


-- 
 Bernát Gábor
Student -  Budapest University of Technology and Economics - Computer
Engineering, M.Sc. - Budapest, Hungary
System Integrator - Gravity R&D | Rock Solid Recommendations - Budapest
office | 5-7 Expo ter | H-1101 | Budapest | Hungary

**Comment by veqryn** [comment]
I would be up for moving the current set of maps out of this repo, and into the svn repo (we'd have to zip them up).  Then having two lists of maps, each of which gets pulled in by the build script using svn commands.  We could also stop having two separate lists, and just go with the larger (extra-maps) list.  

I am against the overhead that would come with maintaining a repo for each map.  I am also against moving the maps away from sourceforge svn unless github can handle at least 3-4gb.

Part of the current set up is historical.  After the issue with H*****, we eliminated all maps except for big world, minimap, capture the flag, and some other dumb one.  After I took over triplea, I slowly added in a few non-A&A maps to the mix, that were decently popular in the lobby, and we also decided to round out TripleA by including at least one map from every important genre that we have pretty good maps for: Napoleonic, NWO, Middle Earth, etc.  We found the only thing really missing was a small size, fast to play, WW2 map, and so Pact of Steel got added.  Pact of Steel 2 became a living documentation of all map features and xml options.
I became more involved with A&A.org, and basically had a request from them that people couldn't figure out how to download maps, and which maps to download.  They just wanted to install TripleA one time and be done with it, with it having all the A&A maps included.  I was a little worried about this, but gave it a go by creating the second build, called 'extra_maps', and not making it the default download.  A few releases later and it became the default because people still couldn't figure out how to download it (they were just clicking the default link).
It make sense to eliminate the non-extra-maps version, since I am no longer worried about the consequences.  H*****'s case against us was always a load of crap, with no merit in any sense.

**Comment by DanVanAtta** [comment]
Thank you sincerely for the detailed response and context.

Github releases only has a limit on the size of on an individual file, but no aggregate size limit. All our map zips should always easily be less than 1GB zipped. We can then put all maps into github releases as a zip, today without issue. We would then get to benefit from githubs REST API and other features. This is very appealing since we could scrape this page to get a map listing. I can see room for generic extension so people can create their own git forks, upload maps to releases, then anyone can point their tripleA game to that github releases to download maps. Could be very good for sharing and testing : )

I think submodules as @gaborbernat pointed out is a viable way to not only get the zips into github releases, but also the map source into git as well. 1 submodule per map is probably overkill, but looking at submodules, looks like a dozen of them would be easy to maintaine, and we get improved version control and most importantly ability to have pull requests to to allow for more automated update of maps.


I can see the desire for a single download to skip the necessity to dowload. I think a lot of that is due to the relatively quick nature of the download. For what it is, it is impressive, but I don't think the game ever really designed well how it would look when maps were not only stored in the application folder, but now were also in a users home folder. Then whether and how it could pick up from a zipped or not-zipped file, and if you had to unzip them, it was/is confusing.


Okay, before wrapping this up and talking about some concrete actions, as devil's advocate: Okay - if we bundle all maps with tripleA, or all the good ones people would play, why do we bother with map downloads at all? In this case only thing we would store in a users home folder would be game saves, a somewhat confusing feature and related code is removed, and a bunch of infrastructure to support it can also go away. If someone really wants the download feature to test out that one map, then they can easily post it somewhere and it's not hard to drop it in the maps folder of your game engine. If the map is good then we'll include it with the game on the next release. 

Personally, I want to keep download so we can have a small engine file.


But, I think a lot of our map problems go away if we store them all in the users home folder. Essentially treat the game as two entities, an engine plus a maps pack. The first pack would have all maps, then others could be the popular ones, or what not, and the packs perhaps overlap. Point being the game engine lives in on windows program files, and all maps are always in ~/triplea/maps (or some other single custom path the user can specify in settings).

In this case the maps survive between game engine updates. This is good since what needs to change is changed. Maps can be also be updated in part or bulk pretty easily. The download functionality still works perfectly as well. With automated packaging, the updates to installer to place maps in a different location *should* be feasible.

Okay, so to summarize and to have some concrete actions proposed:
1. Host map zips in github releases instead of SVN. I'm interested in updating the downloader to support DL from github releases. I'm also interested in getting the initial move of zip files accomplished.

2. I think submodules is a viable way to handle the maps in git, I'd like to continue discussing and probably pursue it a bit in the future.

3. Map organization on deployed systems is a bit scattered, since it is a mix between what has been downloaded and what is in the current game engine (which may or may not be an overlapping set). To simplify this, we install maps into the home folder location, and also download maps to that one location too.

**Comment by veqryn** [comment]
Are github releases versioned?
If someone wanted to pull down all of them in 1 go, but only the latest versions of each, is that easy to do?
I would like to see something else handled by github releases first, like the 1807 release after we get it sorted out, before we make any decision on where maps live.

I actually do not like the idea of someone forking the releases and pointing their triplea towards that.  Having a single source of truth for the current state of maps is one of the nice things that has happened in the last 5 years.

How are git submodules any better than having all the maps in svn, and pulling them in during the build script?

There is also one other issue to deal with that I did not think of fully in my last post: the pact of steel 2 xml is a living documentation of all map options and features (specifically, how to use them, with complete working examples).  If we move that out of the main triplea git repo, I am very afraid that it will stop getting updated.  And no, I do not want to turn this into a wiki.

Last, I would be fine with having all maps get installed to the user triplea maps directory, so long as we still bundled maps with triplea.  There are problems with this though: what happens when you already have that map zip in your directory from an earlier install?  We would need to read and figure out if that map was out of date or not, before we overwrote it (or before we asked the user if we wanted to overwrite, that way we can also say which version the old one is vs the new one).  Furthermore, what happens when the zip is a custom modification you've been working on, of that map?  This is one reason they are split to begin with: so you can work on maps in your user directory.  This all would have to happen during the installation, and I am not sure our installers are this capable.  And what would happen for the all_platforms zip, since it has no installer?

**Comment by DanVanAtta** [comment]
github releases are versioned by tag. We could have as effective of versioning as we have today in SVN, but that would be enforced by process. Adding submodules to the mix though would give us a lot more. We would have individual files, and maintenance of the zips could be automated. So I think that is a goal worth going for. First moving to releases is a good initial increment IMO.

To allow for the "all_maps.zip", a build process would build each zip individually and in aggregate. Those artifacts would be created automatically.

w.r.t single source of truth. That should not be a concern. Using forks to DL would only be something for map editors and for ad-hoc sharing. Theoretically possible today if you can craft a map hosting XML. So effectively the idea is to move into a git backed version. The game engine would ship pointing to the main releases repository URL and not have any forks.

One important feature though is that we could drop the game listing XML. So instead of crafting a custom XML to host maps, users would instead simply post their map to github releases, and could then point the game engine to that URL for download. This would help map makers, they could keep their on-going work in git while still developing. Before they need to create their SVN repo on their own initiative, enough of a pain I'd be surprised if any have gone to those length.

w.r.t bundling maps and home folder. The installation is not quite as nice for users perhaps, but it sets up for a clear and more straight forward map process. One key is that we'll have a game setting for the map folder location. Generally this setting should be correct, but if no maps are found, we'll prompt the user to update the setting to point to where maps are installed.

Then in the all_platforms zips we simply create a second top level folder, triplea_maps that sits next to the versioned folder. Users would expand the zip, drop one folder in one spot, and drop the maps either in the same spot or somewhere else.  Then maps and game engine would be both bundled individually and together.

So if a person knows they have maps, they have options to keep them around. Otherwise an overwrite can be good since that is a full update of all maps in a folder we generally cannot do a full update. (It is possible to redownload all out of date maps, but the process is not streamlined).

There are more features revolving around maps that I'd like to do, making the overall experience better, but with the split map location most are impossible/really difficult. Most of the current UX problems today I believe are artifacts from confusion of the dual location. Once that is resolved we can improve the rest of the UX IMO.

**Comment by veqryn** [comment]
I do not understand the bit about 'pointing to a release'.  

And if we did move to git submodules, we wouldn't actually need github releases at all.  You would just pull in the submodule and package it into a zip, as part of the build process.  

Also, if we used releases, then the links to download from would change each time, instead of being stable (like how a file server has stable links).

I also do not understand how this would eliminate the download xml.  Where would you store the meta information, in a single place, without having to gather it from inside of a submodule or from inside of a zip?

I do not like the idea of packaging the maps in a separate directory and relying on the user to move them.  You seem to think the users are computer literate...

Lastly, I fail to see how having map zips in two places affects the UI or UX in any significant way.
From the user perspective, they all show up on the Choose Games list.
Having the maps in both places buys us a lot actually: We ensure that when you download triplea, you get X common starting maps no matter what, and you can download more.

**Comment by DanVanAtta** [comment]
Pointing to a release means that instead of using this URL: `https://github.com/triplea-game/triplea/releases`, you would use whatever you wanted.

With github submodules or not, the maps would be built into a zip and then pushed into releases.

With releases the links to the download do not change, artifacts are identified by an ID.  There is an easy REST api github provides that we can use. Using this API the game could even present options to download multiple versions of maps along with  some other interesting statistics such as who uploaded maps and more.

The REST API returns a JSON with the information we would need, so the download XML does go away. Pushing artifacts into releases is easy, and since you don't have to do any infrastructure work, it becomes very easy to custom host maps.


I agree in not particularly like having two directories. With one directory you are a folder from a compressed zip files anyways, so worst case the maps and engine are unzipped to the same folder. It's truly not that big of a deal. What's more difficult is finding where the maps are. In general gamers are relatively familiar with the concept of a map packs, this is pretty similar so I don't think will be too confusing.

The two places is significant:
- if someone downloads a map that comes with the game engine, then there is a collision that we coudl otherwise avoid
- there is no easy way to update all downloaded maps
- requiring maps to come with the game engine bloats the size of the game engine by orders of magnitude
- having the maps in the one spot would help for the packaging and simpler code. No longer have to worry about path precedence.
- one search path as a setting for maps is simple and clear for users. Having multiple makes things get more complicated, harder to manipulate. It is also confusing to users whether maps are coming from their folder, the game engine, or old game engineers, or which. I for example was not even aware that there *were* some maps in my home folder, I thought they were all with the game.  So the multi-folder merge is confusing. 


For the record too, I never mentioned UI, just UX, so the fact that the lists are merged into a single UI is a bit immaterial:
> "I fail to see how having map zips in two places affects the UI or UX in any significant way."
If anything this is part of the confusion, it's not obvious where TripleA is finding the maps.


Fundamentally the second location of maps is useful since we can then upgrade and change the engine independently of maps. Also this creates a clear place for maps, that simplifies things. For example, if someone downloads tic tac toe, we would never be able to clean that out. With the second location and the ability to bulk update it, we would have more opportunity.

**Comment by DanVanAtta** [comment]
Okay, two map location decision I don't think was necessarily the best, but it's easier to improve at this point than anything else. 

Considering not everyone in the world has broadband, let's maintain the small vs large download. Since it'll be automated, there should not be so much debate on this. Let's also begin by posting artifacts to github release. I was able to get 20x the DL speed from there than compared to source forge (6MB from github release). With that DL speed, my concern of upgrade pain is much lessoneed.

@veqryn There seems to be confusion about github releases. Please speak up if this is not your understanding as well, but we will be pushing the windows installer, DMG, and all platforms zip to releases.

Concretely, some game updates to help the two location situation:
- when a user has a map in their home folder that comes with the game engine, we ask the user if they would like to remove the unnecessary map. This avoids collisions.
- disable downloads of maps that came with the game 
- add a "update all" feature to the downloader for all maps that were downloaded

Unresolved are some headaches the two locations cause to test, and also the fact the code is not tested which lends to the overall brittle problem, but can't solve every problem all the time..

**Comment by DanVanAtta** [comment]
Addendum, maybe instead of
> "disable downloads of maps that came with the game"

We could have maps that came with the game get downloaded into the correct folder to overwrite the version that came with the game.

**Comment by DanVanAtta** [comment]
There is agreement in this issue that "maps" in the triplea folder can be obtained from somewhere else and that we do not have them in with the source code. This would help map packaging since we woudl download all maps from external rather than some of them and then merging and then packaging. Closing since this was mainly to discuss options. Whoever updates packaging (with notice to: @gaborbernat) is free to make this kind of update.
